### PR TITLE
Use more robust hexadecimal escaping of interpreter configuration.

### LINKED ISSUE #2298: \n in path causing issues on windows
### Bug Description

When building a brand new project created with `maturin new`, I see that error: `failed to parse PyO3 config: Error { value: "expected key=value pair on line 7", source: None }`.

After some investigations, I found out that `from_cargo_dep_env()` replaces every `\\n` with `\n`. https://github.com/PyO3/pyo3/blob/cb485576016f5bcb87e395c78adbbfd09409edad/pyo3-build-config/src/impl_.rs#L369

Here is what `buf.as_str()` looks like before that replace is done: 

> ```"implementation=CPython\\nversion=3.9\\nshared=true\\nabi3=false\\nlib_name=python39\\nlib_dir=C:\\Users\\nobsclement\\AppData\\Local\\Programs\\Python\\Python39\\libs\\nexecutable=D:\\code\\nobody\\rust\\libs\\test_package\\.venv\\Scripts\\python.exe\\npointer_width=64\\nbuild_flags=\\nsuppress_build_script_link_lines=false\\n"```

And after the replace:

> ```"implementation=CPython\nversion=3.9\nshared=true\nabi3=false\nlib_name=python39\nlib_dir=C:\\Users\nobsclement\\AppData\\Local\\Programs\\Python\\Python39\\libs\nexecutable=D:\\code\nobody\\rust\\libs\\test_package\\.venv\\Scripts\\python.exe\npointer_width=64\nbuild_flags=\nsuppress_build_script_link_lines=false\n"```

Note that `lib_dir` and `executable` now have wrong paths: `C:\\Users\nobsclement\\AppData\\Local\\Programs\\Python\\Python39\\libs` and `D:\\code\nobody\\rust\\libs\\test_package\\.venv\\Scripts\\python.exe`.

The bug occurs when the `parse_value` macro trims the line and interprets the path as two (or more) lines:
https://github.com/PyO3/pyo3/blob/cb485576016f5bcb87e395c78adbbfd09409edad/pyo3-build-config/src/impl_.rs#L379

### Steps to Reproduce

On windows:
1. Source a venv in a path that contains a folder that starts with a lowercase n
2. Use `maturin build`

The build crashed when parsing `lib_dir`, but it should also crash when parsing `executable` which those steps should reproduce.

### Backtrace

```shell
error: failed to run custom build command for `pyo3 v0.16.3`

Caused by:
  process didn't exit successfully: `D:\code\nobody\rust\target\debug\build\pyo3-b6ef0b8efd5a0d05\build-script-build` (exit code: 101)
  --- stderr
  thread 'main' panicked at 'failed to parse PyO3 config: Error { value: "expected key=value pair on line 7", source: None }', C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\pyo3-build-config-0.16.3\src\lib.rs:94:10
  stack backtrace:
     0: std::panicking::begin_panic_handler
               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c\/library\std\src\panicking.rs:584
     1: core::panicking::panic_fmt
               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c\/library\core\src\panicking.rs:143
     2: core::result::unwrap_failed
               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c\/library\core\src\result.rs:1749
     3: enum$<core::result::Result<pyo3_build_config::impl_::InterpreterConfig,pyo3_build_config::errors::Error> >::expect<pyo3_build_config::impl_::InterpreterConfig,pyo3_build_config::errors::Error>
               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c\library\core\src\result.rs:1022
     4: pyo3_build_config::get::closure$0
               at C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\pyo3-build-config-0.16.3\src\lib.rs:83
     5: once_cell::sync::impl$6::get_or_init::closure$0<pyo3_build_config::impl_::InterpreterConfig,pyo3_build_config::get::closure_env$0>
               at C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\once_cell-1.10.0\src\lib.rs:974
     6: once_cell::imp::impl$4::initialize::closure$0<pyo3_build_config::impl_::InterpreterConfig,once_cell::sync::impl$6::get_or_init::closure_env$0<pyo3_build_config::impl_::InterpreterConfig,pyo3_build_config::get::closure_env$0>,enum$<once_cell::sync::impl$6:
               at C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\once_cell-1.10.0\src\imp_std.rs:95
     7: once_cell::imp::initialize_inner
               at C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\once_cell-1.10.0\src\imp_std.rs:171
     8: once_cell::imp::OnceCell<pyo3_build_config::impl_::InterpreterConfig>::initialize<pyo3_build_config::impl_::InterpreterConfig,once_cell::sync::impl$6::get_or_init::closure_env$0<pyo3_build_config::impl_::InterpreterConfig,pyo3_build_config::get::closure_e
               at C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\once_cell-1.10.0\src\imp_std.rs:93
     9: once_cell::sync::OnceCell<pyo3_build_config::impl_::InterpreterConfig>::get_or_try_init<pyo3_build_config::impl_::InterpreterConfig,once_cell::sync::impl$6::get_or_init::closure_env$0<pyo3_build_config::impl_::InterpreterConfig,pyo3_build_config::get::clo
               at C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\once_cell-1.10.0\src\lib.rs:1014
    10: once_cell::sync::OnceCell<pyo3_build_config::impl_::InterpreterConfig>::get_or_init<pyo3_build_config::impl_::InterpreterConfig,pyo3_build_config::get::closure_env$0>
               at C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\once_cell-1.10.0\src\lib.rs:974
    11: pyo3_build_config::get
               at C:\Users\nobsclement\.cargo\registry\src\github.com-1ecc6299db9ec823\pyo3-build-config-0.16.3\src\lib.rs:75
    12: build_script_build::configure_pyo3
               at .\build.rs:43
    13: build_script_build::main
               at .\build.rs:56
    14: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c\library\core\src\ops\function.rs:227
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
warning: build failed, waiting for other jobs to finish...
```

### Your operating system and version

Windows 10, Version 2004

### Your Python version (`python --version`)

Python 3.9.10

### Your Rust version (`rustc --version`)

rustc 1.60.0 (7737e0b5c 2022-04-04)

### Your PyO3 version

0.16.3

### How did you install python? Did you use a virtualenv?

Python was installed using the python.org installer for windows.

I used a venv to build my project.

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by adamreichold**
I do feel sorry for adding a hex codec to `pyo3-build-config`. :crying_cat_face:  

Closes #2298

### CODE REVIEW

**Comment by adamreichold** [comment]
@cduranti Just to verify that I understood the issue correctly, would it be possible for you to verify that this works in your setting using a Git dependency? Thank you!

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - Fix segfault when calling FFI methods `PyDateTime_DATE_GET_TZINFO` or `PyDateTime_TIME_GET_TZINFO` on `datetime` or `time` without a tzinfo. [#2289](https://github.com/PyO3/pyo3/pull/2289)
+- Fix directory names starting with the letter n breaking serialization of the interpreter configuration on Windows. [#2299](https://github.com/PyO3/pyo3/pull/2299)
```
Comment (davidhewitt) [inline CHANGELOG.md:24]: Potentially worth mentioning this was a regression from 0.16.3?

**Review by davidhewitt — APPROVED** [review]
Yeah, adding hex is unfortunate, however probably the simplest way to resolve this. Sorry for not catching this bug in the original review, and thank you for fixing!

**Comment by Sookhaal** [comment]
> @cduranti Just to verify that I understood the issue correctly, would it be possible for you to verify that this works in your setting using a Git dependency? Thank you!

Fix is working on my side, thank you very much for being so quick!

**Pull request merged**